### PR TITLE
Wine check & Windows popup helper

### LINKED ISSUE #24: Tell wine users to use the native version.
Just check if HKEY_LOCAL_MACHINE\Software\Wine exists and if it does inform the user we have a native version of the game.

### PR DESCRIPTION
**Pull request by VasilisThePikachu**
Resolves #24

People... do this? Welp one issue down

### CODE REVIEW

File: SS14.Launcher/Program.cs [hunk 1]
```diff
@@ -201,6 +204,30 @@ private static unsafe void CheckBadAntivirus()
         }
     }
 
+    private static unsafe void CheckWine()
+    {
+        if (OperatingSystem.IsWindows())
```
Comment (PJB3005) [inline SS14.Launcher/Program.cs:209]: Invert this if statement with an early return.

File: SS14.Launcher/Program.cs [hunk 2]
```diff
@@ -201,6 +204,30 @@ private static unsafe void CheckBadAntivirus()
...
+            try
+            {
+                Registry.LocalMachine.OpenSubKey(@"SOFTWARE\Wine", true);
```
Comment (PJB3005) [inline SS14.Launcher/Program.cs:212]: You're not closing this registry key in case you successfully open it.

File: SS14.Launcher/Program.cs [hunk 3]
```diff
@@ -201,6 +204,30 @@ private static unsafe void CheckBadAntivirus()
...
+                var text =
+                    $"You seem to be running the launcher under Wine.\n\nWe recommend you run the native linux version instead.";
+                var caption = $"Wine detected!";
+
+                fixed (char* pText = text)
+                fixed (char* pCaption = caption)
+                {
+                    _ = Windows.MessageBoxW(HWND.NULL, (ushort*)pText, (ushort*)pCaption,
+                        MB.MB_OK | MB.MB_ICONWARNING);
+                }
```
Comment (PJB3005) [inline SS14.Launcher/Program.cs:227]: Make this a helper function.

File: SS14.Launcher/Program.cs (hunk 2 above)
Comment (PJB3005) [reply SS14.Launcher/Program.cs:212]: Also surely you wouldn't get a security error if you don't open the registry key as writable? Sounds like a saner way to check this.

**Review by PJB3005 — CHANGES_REQUESTED** [review]

File: SS14.Launcher/Program.cs (hunk 2 above)
Comment (VasilisThePikachu) [reply SS14.Launcher/Program.cs:212]: My dumbass forgot accidentally set it true

File: SS14.Launcher/Program.cs [hunk 4]
```diff
@@ -201,6 +204,28 @@ private static unsafe void CheckBadAntivirus()
...
+    {
+        if (!OperatingSystem.IsWindows())
+            return;
+
+        using var key = Registry.LocalMachine.OpenSubKey(@"SOFTWARE\Wine", false);
+
+        if (key != null)
+        {
+            Log.Debug("Wine detected, showing warning.");
+            var text = $"You seem to be running the launcher under Wine.\n\nWe recommend you run the native linux version instead.";
```
Comment (Visne) [inline SS14.Launcher/Program.cs:217]:
```suggestion
            var text = $"You seem to be running the launcher under Wine.\n\nWe recommend you run the native Linux version instead.";
```

**Review by Visne — CHANGES_REQUESTED** [review]
You should be able to press something like "Don't show again". There might be a valid reason why someone would want to use the Wine/Proton version over native.

**Comment by VasilisThePikachu** [comment]
> You should be able to press something like "Don't show again". There might be a valid reason why someone would want to use the Wine/Proton version over native.

What kind of situation would someone prefer wine over native?

**Comment by deathride58** [comment]
Compatibility. BSD in particular is pretty shonky when it comes to Linux compatibility. Additionally, it's pretty common within the Linux ecosystem for some programs to have weird compatibility issues, especially on more obscure distros, so seeing how crossplatform applications run under Wine ends up being a *very* common troubleshooting step among Linux users.

If someone's running SS14 under Wine, then it's safe to assume they probably know what they're doing.

**Comment by VasilisThePikachu** [comment]
> You should be able to press something like "Don't show again". There might be a valid reason why someone would want to use the Wine/Proton version over native.

I don't think you can do this with the Windows prompt ui (can't supply custom text in the buttons)

**Comment by VasilisThePikachu** [comment]
Could just change the popup to be a yes-no or show one after

**Comment by VasilisThePikachu** [comment]
Just pressing ok makes further warnings not appear, thinks its fine

File: SS14.Launcher/Helpers.cs [hunk 5]
```diff
@@ -171,4 +171,15 @@ public static void ChmodPlusX(string path)
...
+    public static unsafe int MessageBoxHelper(string text, string caption, uint type)
+    {
+        int ret;
+
+        fixed (char* pText = text)
+        fixed (char* pCaption = caption)
+
+            ret = Windows.MessageBoxW(HWND.NULL, (ushort*)pText, (ushort*)pCaption, type);
+
+        return ret;
```
Comment (Visne) [inline SS14.Launcher/Helpers.cs:183]:
```suggestion
        fixed (char* pText = text)
        fixed (char* pCaption = caption)

        return Windows.MessageBoxW(HWND.NULL, (ushort*)pText, (ushort*)pCaption, type);
```

File: SS14.Launcher/Helpers.cs [hunk 6]
```diff
@@ -171,4 +171,11 @@ public static void ChmodPlusX(string path)
             FileAccessPermissions.UserExecute | FileAccessPermissions.GroupExecute |
             FileAccessPermissions.OtherExecute;
     }
+    public static unsafe int MessageBoxHelper(string text, string caption, uint type)
+    {
+        fixed (char* pText = text)
+        fixed (char* pCaption = caption)
+
+        return Windows.MessageBoxW(HWND.NULL, (ushort*)pText, (ushort*)pCaption, type);
+    }
```
Comment (PJB3005) [inline SS14.Launcher/Helpers.cs:180]: Absolutely do not:
1. Use a braceless block for a `fixed()` block.
2. Un-indent a braceless block.

**Review by PJB3005 — CHANGES_REQUESTED** [review]

**Review by Visne — APPROVED** [review]
I don't know anything about unsafe C# but the rest looks fine to me.

**Review by PJB3005 — APPROVED** [review]

**Pull request merged**